### PR TITLE
Implemented simple editor in React

### PR DESCRIPTION
**Pull request by ICimpoes**
Implemented simple editor using React components.
Available actions:
1. Drag a shape from palette to canvas
2. Move shapes in canvas

The current version is deployed on https://icimpoes.github.io/react-train/

### CODE REVIEW

**Comment by ICimpoes** [comment]
The current version is deployed on https://icimpoes.github.io/react-train/

File: src/App.tsx [hunk 1]
```diff
@@ -1,53 +1,9 @@
...
-                    <svg
-                        className="canvas"
-                        viewBox="0 0 1200 600"
-                        version="1.1"
-                        xmlns="http://www.w3.org/2000/svg"
-                    ></svg>
-                </div>
-            </div>
-        );
+        return <Editor />;
```
Comment (ardevelop) [inline src/App.tsx]: Probably a redundant component. Is there anything we want to delegate to this component in time?

File: src/Editor.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import { ShapeType } from "./Shapes";
+import Palette from "./Palette";
+import Workspace from "./Workspace";
+
+export default function Editor() {
+    // symply having useState<ShapeType | undefined> does not work.
+    // when setting state with shape, the ShapeType function is called? Why?
+    const [dragShape, setDragShape] = useState<[ShapeType | undefined]>([
+        undefined,
+    ]);
```
Comment (ardevelop) [inline src/Editor.tsx]: Let's revise if we can avoid using functions as a state value, as it is not a common practice in React

File: src/Editor.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+    return (
+        <div data-testid="editor" className="editor">
+            <Palette
+                setDragShape={(s: ShapeType) => {
```
Comment (ardevelop) [inline src/Editor.tsx]: - Follow the naming convention of html and React for registering callbacks: e.g.: `onStartDrag`
- We should prefer using `React.useCallback` when define callbacks

```tsx
const handleDragStart = React.useCallback((shapeType: ShapeType) => {
  setDragShape([shapeType]);
}, [])

return <Palette onDragStart={handleDragStart} />
```

File: src/Palette.tsx [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import React from "react";
+import Shape, { ShapeType, Shapes } from "./Shapes";
+
+export default function Palette(props: {
+    setDragShape: (shape: ShapeType) => void;
```
Comment (ardevelop) [inline src/Palette.tsx]: For better reading you can extract the props contract as an interface

File: src/Palette.tsx [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+}) {
+    return (
+        <div data-testid="palette" className="palette">
+            {Shapes.map((s, i) => (
+                <Shape key={i} shape={s} setDragShape={props.setDragShape} />
```
Comment (ardevelop) [inline src/Palette.tsx]: We should prefer not using index as key, probably for this case the shape type name is a better option

File: src/Shapes.tsx [hunk 6]
```diff
@@ -0,0 +1,121 @@
+import React, { MouseEvent } from "react";
+
+function circle(
+    x: number,
+    y: number,
+    setDragItem?: (e: Draggable) => void
+): React.JSX.Element {
```
Comment (ardevelop) [inline src/Shapes.tsx]: Probably we want to convert this factories into components so we can use hooks inside

File: src/App.tsx (hunk 1 above)
Comment (ICimpoes) [reply src/App.tsx]: removed the editor component, left only App.

File: src/Editor.tsx (hunk 2 above)
Comment (ICimpoes) [reply src/Editor.tsx]: Done, now ShapeType is `"circle" | "square" | "triangle"`

File: src/Editor.tsx (hunk 3 above)
Comment (ICimpoes) [reply src/Editor.tsx]: updated names and wrapped functions into useCallback

File: src/Shapes.tsx (hunk 6 above)
Comment (ICimpoes) [reply src/Shapes.tsx]: Done, now the shapes are React components

File: src/App.tsx [hunk 7]
```diff
@@ -1,53 +1,15 @@
...
-                            height="50"
-                            version="1.1"
-                            xmlns="http://www.w3.org/2000/svg"
-                        >
-                            <use xlinkHref="assets/shapes.svg#triangle" />
-                        </svg>
-                    </div>
-                </div>
+export default function App() {
+    const [dragShape, setDragShape] = useState<ShapeType | undefined>();
```
Comment (ardevelop) [inline src/App.tsx]: I guess `| undefined` is redundant, it will behave same without it

**Review by ardevelop — APPROVED** [review]

**Pull request merged**